server/src/main/java/org/elasticsearch/action/index/IndexRequest.java

```diff
@@ -717,13 +718,29 @@ public void writeThin(StreamOutput out) throws IOException {
         writeBody(out);
     }
 
+    public void serializeThin(BytesStream out, SerializationContext result) throws IOException {
+        out.writeByte((byte) 0);
+        super.writeThin(out);
+        writeBodyStart(out);
+        result.insertBytesReference(source);
```

Comment:
It would make some sense to consider not doing this below a certain size for `source` (the same is true for slicing of e.g. `SearchHit` sources on the inbound path) when the length of the source is so small that the overhead we introduce from the slicing eats up memory savings.
The cutoff here is above 50 bytes, didn't do the exact math but including the linked list entry in the write-throttling-handler and such, I would say it's rather over 100b than under, so for some use-cases this is a large fraction of the source values I think.